### DISCUSSION TITLE
zh/nixos-with-flakes/start-using-home-manager

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#99] [Announcements]
# zh/nixos-with-flakes/start-using-home-manager

一份非官方的新手指南

https://nixos-and-flakes.thiscute.world/zh/nixos-with-flakes/start-using-home-manager

**Comment by xBLACKICEx** [comment]
`home.file` 和 `xdg.configFile` 有什么区别？

**Reply by ryan4yin** [reply]
`xdg.configFile` 可以简单理解为 `home.file.".config"` 的别名，可以让你少写点东西。
对于这种你没接触过的 options，你可以直接在 Home Manager 的文档中查到它们的详细介绍，文档不详细也可以直接阅读其源码：

https://nix-community.github.io/home-manager/options.xhtml#opt-xdg.configFile

**Reply by xBLACKICEx** [reply]
谢谢

**Comment by Yesterday3896** [comment]
typo "my-nixos" instead of "nixosSystem"

**Comment by ULeang** [comment]
对于最后"如何以特权身份使用 Home Manager 安装的软件包"这一节,我认为作者的阐述不太准确,根据manpage,sudo命令并不是"临时授予当前用户访问系统资源的特权",而是"以另一位用户的身份执行一条命令",这个"另一位用户"一般都是root,在执行这条命令时也就有了root的权限
执行命令时shell根据PATH寻找对应的命令,由于homemanager分别为每一位用户管理PATH中的软件,root的PATH中无法找到该命令,因此报错;但使用sudo,则是在当前用户的PATH中寻找命令,然后sudo会将effective-user-id切换为root的user-id,以此完成"以另一位用户的身份执行一条命令"的效果
比如,执行sudo whoami会显示root而不是当前用户的名字

**Reply by ryan4yin** [reply]
感谢反馈，已修正这部分内容 https://github.com/ryan4yin/nixos-and-flakes-book/commit/4f141293d52165fc8ad5d05b5eab6dcc5b5845ab

**Comment by RobotDisco** [comment]
Do I have to choose between managing home-manager config _entirely_ via `nixosSystem` or `programs.home-manager.enable = true`? Could I configure both, or will that cause conflicts? (I imagine it would be nice to update and iterate on my home-manager config without wanting to update my system config all the time, but perhaps that would cause inconsistencies and conflicts.)

**Reply by ryan4yin** [reply]
Short anwser: Yes. See https://github.com/nix-community/home-manager/issues/3211

**Comment by Ithde** [comment]
安装 `Home Manager` 那一节，出现报错（因为虚拟式上的复制不出来，所以这里就手打了）：
~~~
error:
        ... while calling the 'seq' builtin

            at /nix/store/61502m9h1jl40j7sxbcyjdm6qi1k7x97-source/lib/modules.nix:322:18:

            321|        options = checked options;
            322|        config = checked (removeAttrs config [ "_modules" ]);
               |
            323|        _modules = checked (config._modules);
        ... while calling the 'throw' builtin
        
            at /nix/store/0q50lbnplnhbzqlb2a6gjrqpwzq1i3fc-source/lib/modules.nix/298:18:

            297|                            ''
            298|                else throw baseMsg
               |
            299|            else null;

        error: The option 'home-manager.UseUserPackages' does not exist. Definition values:
        - In `/nix/store/0q50lbnplnhbzqlb2a6gjrqpwzq1i3f-source/flake.nix`: true
~~~
具体的话是这个https://mjj.today/i/j2v9Dm
在两个虚拟机和一个物理机上都有同样的报错，有大佬知道这是啥原因吗?

**Reply by ExcitedEven** [reply]
你说日志是手打的，如果没打错的话。那问题应该是 `UseUserPackages` 错了，应该是小写 `useUserPackages`。

**Comment by JadeYao-z** [comment]
# 设置鼠标指针大小以及字体 DPI（适用于 4K 显示器）
  xresources.properties = {
    "Xcursor.size" = 16;
    "Xft.dpi" = 172;
  };

真的有用吗？为什么我改变数值，一点变化都没有

**Reply by ryan4yin** [reply]
有很多情况会导致这个参数失效，具体细节我也没完全明白，建议问下 AI.

**Comment by Sunjnn** [comment]
执行 `sudo nixos-rebuild switch` 报错了:

```
       error: The option `home-manager.users.sunjnn.home.homeDirectory' has conflicting definition values:
       - In `/nix/store/ilqr95vggnjqpbh6r0fiz2dw43iljg93-source/flake.nix': "/home/sunjnn"
       - In `/nix/store/dimhhb7nhymgd9f35ajdzf6xkij6rsh8-source/nixos/common.nix': "/var/empty"
       Use `lib.mkForce value` or `lib.mkDefault value` to change the priority on any of these definitions.       error: The option `home-manager.users.sunjnn.home.homeDirectory' has conflicting definition values:
       - In `/nix/store/ilqr95vggnjqpbh6r0fiz2dw43iljg93-source/flake.nix': "/home/sunjnn"
       - In `/nix/store/dimhhb7nhymgd9f35ajdzf6xkij6rsh8-source/nixos/common.nix': "/var/empty"
       Use `lib.mkForce value` or `lib.mkDefault value` to change the priority on any of these definitions.
```

这个要怎么解决呢？

**Reply by ryan4yin** [reply]
用 NixOS 的话要学会关键字搜索，你这个问题我一搜就有结果了: https://discourse.nixos.org/t/probably-some-installation-error-about-home-manager/56517

从这个帖子看貌似是你的 NixOS 配置有问题，要把 `users.users.sunjnn.isNormalUser = true;` 加上。

**Comment by clsty** [comment]
“以 sudo 来运行 home-manager 安装的包”，这个仅适用于 NixOS，大概是因为 NixOS 对 sudo 进行了修改或者有自己的实现。
实测在 Debian 13 下，`sudo <cmd>` 会报 `sudo: <cmd>: command not found`，其中 `<cmd>` 是由 Home Manager 安装的包提供的命令。

**Reply by oevery** [reply]
不是 nixos 可以用 `sudo $(which command)` 或者改 path 和软链接，不过不太优雅就是了。

**Comment by Lychee0829** [comment]
home.nix示例中的git 相关配置感觉可以优化一下（
因为运行nixos-rebuild switch输出提示programs.git.userName(userEmail)换成programs.git.settings.user.name(user.email)了
